pkg/gather/azure/azure.go

```diff
@@ -0,0 +1,289 @@
+package azure
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-10-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-06-01/storage"
+	"github.com/Azure/go-autorest/autorest"
+	azureenv "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	azuresession "github.com/openshift/installer/pkg/asset/installconfig/azure"
+	"github.com/openshift/installer/pkg/gather"
+	"github.com/openshift/installer/pkg/gather/providers"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/openshift/installer/pkg/types/azure"
+)
+
+// Gather holds options for resources we want to gather.
+type Gather struct {
+	subscriptionID        string
+	authorizer            autorest.Authorizer
+	environment           azureenv.Environment
+	cloudName             azure.CloudEnvironment
+	resourceGroupName     string
+	logger                logrus.FieldLogger
+	serialLogBundle       string
+	bootstrap             string
+	masters               []string
+	directory             string
+	virtualMachinesClient compute.VirtualMachinesClient
+	accountsClient        storage.AccountsClient
+}
+
+// New returns a Azure Gather from ClusterMetadata.
+func New(logger logrus.FieldLogger, serialLogBundle string, bootstrap string, masters []string, metadata *types.ClusterMetadata) (providers.Gather, error) {
+	cloudName := metadata.Azure.CloudName
+	if cloudName == "" {
+		cloudName = azure.PublicCloud
+	}
+
+	resourceGroupName := metadata.Azure.ResourceGroupName
+	if resourceGroupName == "" {
+		resourceGroupName = metadata.InfraID + "-rg"
+	}
+
+	session, err := azuresession.GetSession(cloudName, metadata.Azure.ARMEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	gather := &Gather{
+		subscriptionID:    session.Credentials.SubscriptionID,
+		authorizer:        session.Authorizer,
+		environment:       session.Environment,
+		resourceGroupName: resourceGroupName,
+		logger:            logger,
+		serialLogBundle:   serialLogBundle,
+		bootstrap:         bootstrap,
+		masters:           masters,
+		directory:         filepath.Dir(serialLogBundle),
+		cloudName:         cloudName,
+	}
+
+	gather.virtualMachinesClient = compute.NewVirtualMachinesClientWithBaseURI(session.Environment.ResourceManagerEndpoint, gather.subscriptionID)
+	gather.virtualMachinesClient.Authorizer = gather.authorizer
+
+	gather.accountsClient = storage.NewAccountsClientWithBaseURI(session.Environment.ResourceManagerEndpoint, gather.subscriptionID)
+	gather.accountsClient.Authorizer = gather.authorizer
+
+	return gather, nil
+}
+
+// Run is the entrypoint to start the gather process.
+func (g *Gather) Run() error {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```

Comment:
We do this all over the installer. I've never bothered to look it up, but if it is indeed an issue, it's something that needs to be done in a lot more places. I follow this pattern because it's what has been in the installer since I began working on it.

Reply:
> We do this all over the installer. I've never bothered to look it up, but if it is indeed an issue, it's something that needs to be done in a lot more places. I follow this pattern because it's what has been in the installer since I began working on it.

Let's make sure we're talking about the same issue. I should have been more clear to say "this context timeout is being shared." Typically in the installer, we pass a background or TODO context until we're actually making the API call. That is fine.

What I'm pointing out is that in this code I think you're setting a 1 minute deadline for 3 subsequent API calls. I don't think this pattern is common in the installer, but if it is, we can examine it.

I have put together a go playground to illustrate the problem here: https://go.dev/play/p/974k2h_rCEl

Reply:
Fixed.